core/server/master/src/main/java/alluxio/master/scheduler/Scheduler.java

```diff
@@ -0,0 +1,390 @@
+import alluxio.wire.WorkerInfo;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * The Scheduler which controls jobs. It's not thread-safe since start and stop
+ * method is not thread-safe. But we should only have one thread call these two method.
+ */
+@ThreadSafe
+public final class Scheduler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Scheduler.class);
+  private static final int CAPACITY = 100;
+  private static final long WORKER_UPDATE_INTERVAL = Configuration.getMs(
+      PropertyKey.MASTER_WORKER_INFO_CACHE_REFRESH_TIME);
+  private static final int EXECUTOR_SHUTDOWN_MS = 10 * Constants.SECOND_MS;
+  private final Map<JobDescription, Job<?>>
+      mExistingJobs = new ConcurrentHashMap<>();
+  private final Map<Job<?>, Set<WorkerInfo>> mRunningTasks = new ConcurrentHashMap<>();
+  private final JobMetaStore mJobMetaStore;
+  // initial thread in start method since we would stop and start thread when gainPrimacy
+  private ScheduledExecutorService mSchedulerExecutor;
+  private volatile boolean mRunning = false;
+  private Map<WorkerInfo, CloseableResource<BlockWorkerClient>> mActiveWorkers = ImmutableMap.of();
+  private final WorkerProvider mWorkerProvider;
+
+  /**
+   * Constructor.
+   *
+   * @param workerProvider   workerProvider
+   * @param jobMetaStore     jobMetaStore
+   */
+  public Scheduler(WorkerProvider workerProvider, JobMetaStore jobMetaStore) {
+    mWorkerProvider = workerProvider;
+    mJobMetaStore = jobMetaStore;
+  }
+
+  /**
+   * Start scheduler.
+   */
+  public void start() {
+    if (!mRunning) {
+      retrieveJobs();
+      mSchedulerExecutor = Executors.newSingleThreadScheduledExecutor(
+          ThreadFactoryUtils.build("scheduler", false));
+      mSchedulerExecutor.scheduleAtFixedRate(this::updateWorkers, 0, WORKER_UPDATE_INTERVAL,
+          TimeUnit.MILLISECONDS);
+      mSchedulerExecutor.scheduleWithFixedDelay(this::processJobs, 0, 100, TimeUnit.MILLISECONDS);
+      mSchedulerExecutor.scheduleWithFixedDelay(this::cleanupStaleJob, 1, 1, TimeUnit.HOURS);
+      mRunning = true;
+    }
+  }
+
+  private void retrieveJobs() {
+    for (Job<?> job : mJobMetaStore.getJobs()) {
+      mExistingJobs.put(job.getDescription(), job);
+      if (job.isDone()) {
+        mRunningTasks.remove(job);
+      }
+      else {
+        mRunningTasks.put(job, new HashSet<>());
+      }
+    }
+  }
+
+  /**
+   * Stop scheduler.
+   */
+  public void stop() {
+    if (mRunning) {
```

Comment:
@tcrain so one thing I want to verify is that we stop rpc server before calling this scheduler.stop so we can safely cleanup the state